base/acme/src/main/java/org/dogtagpki/acme/server/ACMEChallengeProcessor.java

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright Red Hat, Inc.
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+//
+package org.dogtagpki.acme.server;
+
+import java.util.Collection;
+import java.util.Date;
+
+import org.dogtagpki.acme.ACMEAccount;
+import org.dogtagpki.acme.ACMEAuthorization;
+import org.dogtagpki.acme.ACMEChallenge;
+import org.dogtagpki.acme.ACMEOrder;
+import org.dogtagpki.acme.validator.ACMEValidator;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class ACMEChallengeProcessor implements Runnable {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ACMEChallengeProcessor.class);
+
+    ACMEAccount account;
+    ACMEAuthorization authorization;
+    ACMEChallenge challenge;
+    ACMEValidator validator;
+
+    public ACMEChallengeProcessor(
+            ACMEAccount account,
+            ACMEAuthorization authorization,
+            ACMEChallenge challenge,
+            ACMEValidator validator) {
+
+        this.account = account;
+        this.authorization = authorization;
+        this.challenge = challenge;
+        this.validator = validator;
+    }
+
+    public void run() {
+        try {
+            processChallenge();
+        } catch (Exception e) {
+            logger.error("Unable to process challenge " + challenge.getID() + ": " + e.getMessage(), e);
+        }
+    }
+
+    public void processChallenge() throws Exception {
+
+        String authzID = authorization.getID();
+        String challengeID = challenge.getID();
+
+        logger.info("Processing challenge " + challengeID);
+
+        ACMEEngine engine = ACMEEngine.getInstance();
+
+        try {
+            validator.validateChallenge(authorization, challenge);
```

Comment:
The behaviour I discuss in this comment was not introduced by this PR, but because we're making these changes it might be a good time to fix it.

https://tools.ietf.org/html/rfc8555#section-8.2 states:

```
   The server MUST provide information about its retry state to the
   client via the "error" field in the challenge and the Retry-After
   HTTP header field in response to requests to the challenge resource.
   The server MUST add an entry to the "error" field in the challenge
   after each failed validation query.  The server SHOULD set the Retry-
   After header field to a time after the server's next validation
   query, since the status of the challenge will not change until that
   time.
```

So, IMO we need to do something like:

- Extract retry behaviour out of the `ACMEValidator` classes and into `ACMEChallengeProcessor` i.e. at this location.
- Therefore the `ACMEValidator.validateChallenge()` contract becomes a *single attempt*.
- To comply with RFC, `ACMEChallengeProcessor` must store error info for each failed attempt.
- We maybe need to store a "last attempt" or "next attempt" field on the challenge so that we can produce the appropriate `Retry-After` header.  This is a SHOULD not a MUST, but as it involves schema, if we think we want to do it, the sooner the better!

Reply:
Agreed on all, but does it mean `ACMEChallenge.error` will need to be multi-valued? Do we need to store a timestamp associated with each error?